File: src/engines/llamacpp.ts
import { execFileSync } from "node:child_process";
import type { EngineAdapter, StartedServer } from "../types.js";
import { spawnServerAndWaitReady } from "../harness/spawn-server.js";
import { EngineNotFoundError, UsageError } from "../errors.js";

const BINARY = "llama-server";

export class LlamaCppAdapter implements EngineAdapter {
  readonly name = "llama.cpp";
  readonly binary = BINARY;

  isInstalled(): boolean {
    try {
      execFileSync(BINARY, ["--version"], { stdio: "ignore" });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * `model` is treated as a Hugging Face repo spec (e.g.
   * "bartowski/Qwen2.5-1.5B-Instruct-GGUF:Q4_K_M") -- llama.cpp's own
   * -hf flag downloads and caches it automatically, no manual step needed.
   */
  async startServer(opts: {
    model: string;
    port: number;
    timeoutMs: number;
    verbose?: boolean;
  }): Promise<StartedServer> {
    if (!this.isInstalled()) {
      throw new EngineNotFoundError(this.name, BINARY);
    }
    if (opts.model.startsWith("-")) {
      throw new UsageError(
        `Invalid --model value "${opts.model}": cannot start with "-" (would be parsed as a flag by llama-server, not a model spec)`,
      );
    }
    const spawned = await spawnServerAndWaitReady({
      engine: this.name,
      command: BINARY,
      args: [
        "-hf",
        opts.model,
        "--port",
        String(opts.port),
        "--host",
        "127.0.0.1",
      ],
      readyCheckUrl: `http://127.0.0.1:${opts.port}/v1/models`,
      timeoutMs: opts.timeoutMs,
      verbose: opts.verbose,
    });
    return {
      process: spawned.process,
      baseUrl: `http://127.0.0.1:${opts.port}`,
      modelId: "default",
      stop: spawned.stop,
    };
  }
}

File: python/src/inferbench/engines/llamacpp.py
"""
Ported from src/engines/llamacpp.ts. `model` is treated as a Hugging Face
repo spec (e.g. "bartowski/Qwen2.5-1.5B-Instruct-GGUF:Q4_K_M") -- llama.cpp's
own `-hf` flag downloads and caches it automatically, no manual step needed.
"""
from __future__ import annotations

import subprocess

from ..errors import EngineNotFoundError, UsageError
from ..harness.spawn_server import spawn_server_and_wait_ready
from ..types import StartedServer

_BINARY = "llama-server"


class LlamaCppAdapter:
    name = "llama.cpp"
    binary = _BINARY

    def is_installed(self) -> bool:
        try:
            subprocess.run(  # noqa: S603, S607 -- fixed argv, binary name only, no shell
                [_BINARY, "--version"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                check=True,
            )
            return True
        except (OSError, subprocess.CalledProcessError):
            return False

    def start_server(
        self,
        *,
        model: str,
        port: int,
        timeout_ms: int,
        verbose: bool = False,
    ) -> StartedServer:
        if not self.is_installed():
            raise EngineNotFoundError(self.name, _BINARY)
        if model.startswith("-"):
            raise UsageError(
                f'Invalid --model value "{model}": cannot start with "-" '
                "(would be parsed as a flag by llama-server, not a model spec)"
            )
        spawned = spawn_server_and_wait_ready(
            engine=self.name,
            command=_BINARY,
            args=["-hf", model, "--port", str(port), "--host", "127.0.0.1"],
            ready_check_url=f"http://127.0.0.1:{port}/v1/models",
            timeout_ms=timeout_ms,
            verbose=verbose,
        )
        return StartedServer(
            process=spawned.process,
            base_url=f"http://127.0.0.1:{port}",
            model_id="default",
            stop=spawned.stop,
        )
